Sources/NIOHTTP1/HTTPServerPipelineHandler.swift

```diff
@@ -76,26 +76,36 @@ public final class HTTPServerPipelineHandler: ChannelDuplexHandler {
         /// let the next request in.
         case responseEndPending
 
-        /// Neither a request nor a response are complete. This can be either
-        /// because nothing is happening on the connection, or because the request
-        /// and response are simultaneously progressing and have not completed.
+        /// We are in the middle of both a request and a response and waiting for both `.end`s.
+        case requestAndResponseEndPending
+
+        /// Nothing is active on this connection, the next message we expect would be a request `.head`.
         case idle
 
         /// The server has responded early, before the request has completed. We need
         /// to wait for the request to complete, but won't block anything.
         case requestEndPending
 
+        mutating func requestHeadReceived() {
+            switch self {
+            case .idle:
+                self = .requestAndResponseEndPending
+            default:
```

Comment:
Might be better to explicitly enumerate all the states here?

Reply:
Or use an `if case .idle = self`, as it seems just one case is evaluated.

Reply:
@BasThomas we want to crash in the other cases; @Lukasa ok